FDK/Tools/SharedData/FDKScripts/CheckOutlinesUFO.py

```diff
@@ -12,1020 +13,1145 @@
+    # Used to sort contours.
+    max_p = contour[0]
+    for point in contour:
+        if point.segmentType is None:
+            continue
+        if max_p.y < point.y:
+            max_p = point
+        elif max_p.y == point.y:
+            if max_p.x < point.y:
+                max_p = point
+    contour.maxP = max_p
+
+
+def sort_contours(c1, c2):
+    if not hasattr(c1, 'maxP'):
+        set_max_p(c1)
+    if not hasattr(c2, 'maxP'):
+        set_max_p(c2)
+
+    if c1.maxP.y > c2.maxP.y:
+        return 1
+    elif c1.maxP.y < c2.maxP.y:
+        return -1
+
+    if c1.maxP.x > c2.maxP.x:
+        return 1
+    elif c1.maxP.x < c2.maxP.x:
+        return -1
+
+    lc1 = len(c1)
+    lc2 = len(c2)
+    if lc1 > lc2:
+        return 1
+    elif lc1 < lc2:
+        return -1
+    return 0
+
+
+def restore_contour_order(fixed_glyph, original_contours):
+    """ The pyClipper library first sorts all the outlines by x position,
+    then y position. I try to undo that, so that un-touched contours will end
+    up in the same order as the in the original, and any conbined contours
+    will end up in a similar order. The reason I try to match new contours
+    to the old is to reduce arbitraryness in the new contour order between
+    similar fonts. I can't completely avoid this, but I can reduce how often
+    it happens.
+    """
+    new_contours = list(fixed_glyph)
+    if len(new_contours) > 1:
+        new_contours.sort(sort_contours)
+    else:
+        set_max_p(new_contours[0])
+    new_index_list = range(len(new_contours))
+    new_list = [[i, new_contours[i]] for i in new_index_list]
+    old_index_list = range(len(original_contours))
+    old_list = [[i, original_contours[i]] for i in old_index_list]
+    order_list = []
+
+    # Match contours that have not changed.
+    # This will fix the order of the contours that have not been touched.
+    num_contours = len(new_list)
+    if num_contours > 0:  # If the new contours aren't already all matched..
+        new_index_list = range(num_contours)
+        for i in new_index_list:
+            ci, contour = new_list[i]
+            for j in old_index_list:
+                ci2, old_contour = old_list[j]
+                # skip if is not the same set of values.
+                if len(old_contour) != len(contour):
+                    continue
+                if old_contour[0] != old_contour[0]:
+                    continue
+                order_list.append([ci2, ci])
+                new_list[i] = None
+                break
+
+    new_list = filter(lambda entry: entry is not None, new_list)
+    num_contours = len(new_list)
+    # Check each extreme for a match.
+    if num_contours > 0:
+        new_index_list = range(num_contours)
+        for i in new_index_list:
+            ci, contour = new_list[i]
+            max_p = contour.maxP
+            # Now search the old contour list.
+            for j in old_index_list:
+                ci2, old_contour = old_list
```

Comment:
boolean

Reply:
must have hit the wrong line, but I think a boolean is needed nearby

Reply:
OK. I think I got all of the 0/1 to False/Trues in general.